adoc/chapters/programming_interface.adoc

```diff
@@ -5186,718 +5173,637 @@ If no template argument is provided, then the default allocator for the SYCL
 
-a@
-[source]
-----
-template <typename Container>
-buffer(Container& container,
-       const property_list& propList = {})
+'''
+
+.[apititle]#Construct from a container#
+[source,role=synopsis,id=api:buffer-ctor-container]
 ----
-   a@ Construct a one dimensional SYCL [code]#buffer# instance
-      from the elements starting at [code]#std::data(container)#
-      and containing [code]#std::size(container)# number of elements.
-      The buffer is initialized with the contents of [code]#container#,
-      and the buffer assumes exclusive access to [code]#container# for
-      the duration of its lifetime.
+template <typename Container>                                      (1)
+buffer(Container& container, AllocatorT allocator,
+       const property_list& propList = {});
 
-Data is written back to [code]#container# before the completion of
-[code]#buffer# destruction if the return type of [code]#std::data(container)#
-is not [code]#const#.
+template <typename Container>                                      (2)
+buffer(Container& container, const property_list& propList = {});
+----
+Preconditions: [code]#container# is a contiguous container.
 
-The constructed SYCL [code]#buffer# will use a default constructed [code]#AllocatorT# when allocating memory on the host.
+Constraints: Available only when:
 
-Zero or more properties can be provided to the constructed SYCL
-[code]#buffer# via an instance of
-[code]#property_list#.
+* [code]#std::data(container)# and [code]#std::size(container)# are well formed;
 
-This constructor is only defined for a [code]#buffer# parameterized
-with [code]#Dimensions == 1#, and when [code]#std::data(container)#
-is convertible to [code]#T*#.
+* the return type of [code]#std::data(container)# is convertible to [code]#T*#;
+  and
 
+* [code]#Dimensions == 1#.
 
-a@
-[source]
-----
-template <typename Container>
-buffer(Container& container, AllocatorT allocator,
-       const property_list& propList = {})
-----
-   a@ Construct a one dimensional SYCL [code]#buffer# instance
-      from the elements starting at [code]#std::data(container)#
-      and containing [code]#std::size(container)# number of elements.
-      The buffer is initialized with the contents of [code]#container#,
-      and the buffer assumes exclusive access to [code]#container# for
-      the duration of its lifetime.
+_Effects (1):_ Construct a one dimensional SYCL [code]#buffer# instance from the
+elements starting at [code]#std::data(container)# and containing
+[code]#std::size(container)# number of elements.
+The buffer is initialized with the contents of [code]#container#, and the buffer
+assumes exclusive access to [code]#container# for the duration of its lifetime.
 
 Data is written back to [code]#container# before the completion of
-[code]#buffer# destruction if the return type of [code]#std::data(container)#
-is not [code]#const#.
+[code]#buffer# destruction if the return type of [code]#std::data(container)# is
+not [code]#const#.
```

Comment:
I think we should turn this into a _Remark_, because it's not really an effect of the constructor.